src/state.rs

```diff
@@ -30,14 +38,56 @@ pub struct State {
 }
 
 impl State {
+    /// Loads the state.json from the filesystem, waiting for a file lock to ensure no other
+    /// concurrent Kubie processes are accessing/writing the file at the same time.
     pub fn load() -> Result<State> {
+        Self::access(|state| Ok(state))
+    }
+
+    /// Takes a closure that allows for modifications of the state. Automatically handles
+    /// locking/unlocking and saving after execution of the closure.
+    pub fn modify<F: FnOnce(&mut State) -> Result<()>>(func: F) -> Result<()> {
+        Self::access(|mut state| {
+            func(&mut state)?;
+            state.save()?;
+            Ok(())
+        })
+    }
+
+    fn access<R, F: FnOnce(State) -> Result<R>>(func: F) -> Result<R> {
+        let path = KUBIE_STATE_LOCK_PATH.to_str().unwrap();
+
+        // Acquire the lock
+        let flock = File::create(lock_path())?;
+        flock
+            .lock_exclusive()
+            .with_context(|| format!("Failed to lock state: {}", path))?;
+
+        // Do the work
+        let state = State::read_and_parse()
+            .with_context(|| format!("Could not load state file: {}", KUBIE_STATE_PATH.to_str().unwrap()));
```

Comment:
Can you use Path.display here instead of panicking?

Reply:
Good catch! I've cleaned all of those up.